### ISSUE TITLE
Tab bar navigation development tasks

### ISSUE THREAD

**Issue by defagos** [SRGSSR/playsrg-ios#118]
We have provided an early build of the application in which the burger menu has (finally) be replaced with a tab bar. Here is the list of the issues we have found so far in testing this new version:

### Layout and design

- [x] Final tab icons. Larger maybe, as standard Apple app icons are larger.
- [x] Discuss navigation bar titles, the Play logo and channel colors. Proposal: For audio, display _Audios_ on colored background. For video, display _Videos_.
- [x] Under the livestream tab, the cells do not adjust according to their content.
- [x] Move settings button from homepages to the library tab.
- [x] Incorrect scroll content offsets and indicator insets after mini player dismissal.
- [x] Landscape support for phones must be disabled. 
- [x] With a Bluetooth headset apparently, audio glitches when playing another content while the mini player has been paused. Should check whether this problem existed before in production.
- [x] Mini player content not pinned correctly to the left / right safe area layout guides (visible if the phone is held horizontally. Probably not so relevant if we drop landscape support on phones, but might be required for proper iPad Pro support.
- [x] Library screen design (fonts, colors, sizes).
- [x] Margins between swimlanes on homepages?
- [x] Notification red dots not uniform (the one on the tab bar is standard).
- [x] Consistent tab bar titles.
- [x] Livestream section layout for iPad.
- [x] Center tab strip if fitting in its frame.
- [x] Broken list layouts after returning from a landscape player on phones (with locked portrait orientation). Affects most views (topic detail, shows, lives, etc.).
- [x] Couleur 3 radio logo is not consistent with other radio logos on the radio livestreams view.
- [x] ~Notifications: Move red dot to the left, probably align layout vertically (refer to preparatory design work).~ Not in the final design, current top left position is preserved.
- [x] Use icons with appropriate size (e.g. icon on the account header)
- [x] ~Mini player audio / video generic icons validated?~
- [x] Single radio channel / live section: Do not display upper tabs.

### Mini player

- [x] Decide and align mini player behavior for video content and optional video background playback support.
- [x] Avoid displaying the mini player with the full-screen player at the same time.
- [x] Allow Play/Pause on the mini Player when tapping in the pause icon (don't reopen the player page), for all media.
- [x] Layout Google Cast and standard mini players in the same way.
- [x] When a video has been reduced in the mini player, the play button can be used to play it in the background, even if background playback is disabled. Instead, tapping anywhere on the video mini player should open the full screen player when background playback is disabled.
- [x] With background playback disabled, the video mini player is removed if sending the app to the background and returning. It is probably worth checking all use cases too (audio, video, with and without background playback, playing or paused). When launching the app we don’t restore anything for the moment.
- [x] We could probably update the standard mini player with the most recent item, even if it was played with Google Cast (play some content normally, then enable Google Cast. Play something else. The first content is the one sitting in the mini player after disconnecting Google Cast).
- [x] Now that we can enable Google Cast outside the player, we must stop local playback (which might be made with PiP or AirPlay). We must also continue playback with Google Cast if playing locally while connecting to a receiver.
- [x] Mini player stop button for livestreams. Have it stop playback instead of pausing it.

### Features

- [x] Restore last used tab and section (TV, radio channel, livestream).
- [x] Radio livestream quick access buttons on the player are redundant and must be removed.
- [x] Properly respond to settings change (e.g. server). If not easy, display a message before killing the app.
- [x] Open previously selected regional radio when tapping on a media from the livestreams tab. Note that regional radio choice can now only be made from the player.
- [x] Do not display onboarding screens at startup anymore (but keep them as reference in the settings).
- [x] Remove duplicate onboardings.

### Bugs and technical issues

- [x] Crash when closing the mini player while playing some content on iPad. Probably related to a [Letterbox issue](https://github.com/SRGSSR/srgletterbox-apple/issues/202) since affects only iPad.
- [x] [Crash (on iPad with Airpods headphones)](https://appcenter.ms/orgs/RTS-Organization/apps/Play-RTS-Nightly-1/crashes/errors/3089134397u/overview): Play a video with background mode enabled, reduce the player and remove Airpods. The app crashes. Probably related to a [Letterbox issue](https://github.com/SRGSSR/srgletterbox-apple/issues/202) since affects only iPad.
- [x] Audio glitch when tapping on the video mini player (background playback enabled), when the player is restored. Also not possible to play / pause videos in the mini player, the player opens but the button does not react.
- [x] Loading indicator not visible when resuming playback from the mini player (Letterbox issue).
- [x] Perceived sluggishness when opening the application or a content. Real or just an impression?
- [x] Application hangs after a phone call. Reproducible? Does it affect the production version? _Could not reproduce_
- [x] Non-working 3D touch shortcuts.
- [x] Tab strip has erratic behavior when dragging. ~On the livestream page, it does not follow page swiping either, whereas it does elsewhere.~ Not with the Material component we are now using.
- [x] Video stops and resumes when resuming from the mini player. AirPlay only?
- [x] ~New perceived sluggishness when scrolling on some devices.~
- [x] Per channel favorites are not updated when the first favorite is added (open a homepage, add a favorite, go back. The list has not been updated).
- [x] Fix sometimes empty screen when loading data. Better seen with a slow network.
- [x] Save to history at the end of playback or when paused (currently, short completed content restarts near the end). Alternatively, return to 1 second periodicity.
- [x] ~Fix incorrect content insets for the web view controller after player dismissal.~ Works with `WebViewController`.
- [x] ~Autoplay does not chain videos when using AirPlay.~ _Works like a charm_
- [x] When using PiP, tapping on the pause button in the mini player incorrectly opens the full screen player.
- [x] Display first favorite shows in home page without killing the application.
- [x] Fix collection view layout warnings. _All except on search view have been fixed_
- [x] Fix warning for `HomeViewController`:

```
[TableView] Warning once only: UITableView was told to layout its visible cells and other contents without being in the view hierarchy (the table view or one of its superviews has not been added to a window). This may cause bugs by forcing views inside the table view to load and perform layout without accurate information (e.g. table view bounds, trait collection, layout margins, safe area insets, etc), and will also cause unnecessary performance overhead due to extra layout passes. Make a symbolic breakpoint at UITableViewAlertForLayoutOutsideViewHierarchy to catch this in the debugger and see what caused this to occur, so you can avoid this action altogether if possible, or defer it until the table view has been added to a window. Table view: <UITableView: 0x7fd48e8f1000; frame = (0 0; 639 1366); clipsToBounds = YES; autoresize = RM+BM; gestureRecognizers = <NSArray: 0x600002141950>; layer = <CALayer: 0x600002f110a0>; contentOffset: {0, -74}; contentSize: {639, 3020.2001100029775}; adjustedContentInset: {74, 0, 65, 0}; dataSource: <HomeViewController: 0x7fd48d717520; view: <UIView: 0x7fd490309cd0; frame = (0 0; 639 1366); autoresize = W+H; layer = <CALayer: 0x600002f112a0>>; superview: <UIViewControllerWrapperView: 0x7fd490022d20; frame = (0 0; 639 1366); autoresize = W+H; layer = <CALayer: 0x600002ffc660>>>>
```

- [x] Fix warning:

```
[TableView] Setting the background color on UITableViewHeaderFooterView has been deprecated. Please set a custom UIView with your desired background color to the backgroundView property instead.
```

### Accessibility

- [x] Fix VoiceOver support (tabs, mini player, livestream cells).
- [x] Variable font size support.

### Cleanup
 
- [x] Have a single application section for video and radio channel overviews (branching to the correct one is made on the presence of an optional radio channel parameter).
- [x] Remove support for useless remote configuration parameters.
- [x] Remove alternate radio homepage design.
- [x] Deal with TODOs and FIXMEs.
- [x] Move `Home` directory elsewhere in the project tree, split views and controllers into logical groups.
- [x] Eliminate storyboards with no value (e.g. `HomeMediasViewController`). Instantiate the basic stuff in code.
- [x] If radio channel colors are dropped, cleanup `NavigationController` code accordingly.
- [x] Cleanup unused resources (e.g. navigation bar Play logo if not used anymore).

### Documentation

- [x] Update remote configuration documentation.

### Checks

- [x] Check leaks with Instruments.
- [x] Check network activity.
- [x] Check startup performance.

### Questions

- [x] `videoHomeSections` in code, but `videoSections` in remote configuration (same for audio and livestream sections). Which convention should we use? If `videoHome`, should we rename `HomeViewController` as `VideoHomeViewController`, and `AudiosViewController` as `AudioHomeViewController`?
- [x] Do we need to keep section titles (seen with VoiceOver enabled) in the Library section?
- [x] Should we disable autoplay on videos played in background? _I suggest we don’t do anything for the moment. I think having them continue makes sense according to my tests. This is also consistent with audio and does not require further explanation to the user. Of course the user might forget about her device and play video content for a long time, but this is probably more the exception than the rule_
- [x] Page view analytics?
- [x] Display Google Cast button on homepages?
- [x] Are we removing onboardings entirely, or keeping them as documentation in settings?
- [x] Shouldn’t we display the last synchronisation date footer as footer of Downloads, Favorites, etc.? _Discussed. This is really a technical information, best kept in the settings area_

**Comment by defagos** [comment]
After a global code review, here are a few things which remain to be discussed or done:

- `PlayApplicationNavigation`: Maybe we could introduce a `UIViewController` category method which can therefore be applied anywhere, but checks protocol conformance like in `LivestreamsViewController` or `RadioChannelsViewController`.
- `-[HomeViewController openApplicationSectionInfo:]` logic could probably be rewritten in a more straightforward manner.
- Shouldn't `ApplicationSectionShowAZ`or `ApplicationSectionShowByDate` (and maybe similar constants) be removed and replaced with some contextual information for the videos or audios application sections?
- `-[MediaPlayerViewController reloadDetailsWithMedia:mainChapterMedia:` is a mess. We had to fix the code for `media = nil`, but other methods not supporting nullability according to their nullability attributes are still called with `nil`. We can of course ask ourselves if these methods should accept `nil`, but we should break down this huge method into smaller parts (generic information, view count, subtitle / AD) and properly test for `media` and `mainChapterMedia` nullability.

**Comment by defagos** [comment]
I found no particular guidelines about tab bar accessibility. The documentation [officially declares](https://developer.apple.com/documentation/uikit/uitabbar?language=objc) the tab bar as being accessible, probably by searching on screen with the finger.

The [HIG](https://developer.apple.com/design/human-interface-guidelines/accessibility/overview/introduction/) do not mention the tab bar explicitly either.

The current behavior we have is consistent with the behavior observed in other apps (e.g. Apple Clock). I doubt there is anything else to do. I navigated through the app with VoiceOver and noticed no real issue because of the work we did on the tab bar.
